File: intermediate/08dp/05ninja/solutions/ninja.cs
using System;
using System.Linq;

public class Solution
{
    public static void Main()
    {
        var n = int.Parse(Console.ReadLine());
        var lab = new int[n][];
        var result = new int[n][];
        for(var i = 0; i < n; i++)
        {
            result[i] = new int[n];
            lab[i] = Console.ReadLine()
                .Split(' ')
                .Select(int.Parse)
                .ToArray();
        }


        for (int row = 0; row < n; ++row) {
            for (int col = 0; col < n; ++col) {
                var top = 0;
                var left = 0;
                if (row - 1 >= 0) {
                    top = result[row - 1][col];
                    if (lab[row - 1][col] == lab[row][col]) {
                        top++;
                    }
                }

                if (col - 1 >= 0) {
                    left = result[row][col - 1];
                    if (lab[row][col - 1] == lab[row][col]) {
                        left++;
                    }
                }

                if (top != 0 && left != 0) {
                    result[row][col] = Math.Min(top, left) + 1;
                }
                else {
                    result[row][col] = Math.Max(top, left) + 1;
                }
            }
        }

        Console.WriteLine(result[n - 1][n - 1]);
    }
}

File: intermediate/08dp/05ninja/solutions/ninja.py
import sys

def solve():
    size = int(sys.stdin.readline())

    lab = [list(map(int, sys.stdin.readline().split(' '))) for i in range(size)]

    result = [[0] * size for i in range(size)]
    for row in range(size):
        for col in range(size):
            top = 0
            left = 0
            if row - 1 >= 0:
                top = result[row-1][col]

                if lab[row-1][col] == lab[row][col]:
                    top += 1

            if col - 1 >= 0:
                left = result[row][col-1]
                if lab[row][col-1] == lab[row][col]:
                    left += 1

            if top and left:
                result[row][col] = min([top, left]) + 1
            else:
                result[row][col] = max([top, left]) + 1

    print(result[size-1][size-1])

solve()
